src/realm/db.cpp

```diff
@@ -2332,13 +2336,39 @@ TransactionRef DB::start_frozen(VersionID version_id)
 {
     if (!is_attached())
         throw LogicError(LogicError::wrong_transact_state);
+    // prune stale frozen transactions
+    std::lock_guard<std::recursive_mutex> lock(m_mutex);
+    while (!m_frozen_transactions.empty() && m_frozen_transactions.begin()->second.expired()) {
+        m_frozen_transactions.erase(m_frozen_transactions.begin());
+    }
     ReadLockInfo read_lock;
-    grab_read_lock(read_lock, version_id);
     ReadLockGuard g(*this, read_lock);
-    Transaction* tr = new Transaction(shared_from_this(), &m_alloc, read_lock, DB::transact_Frozen);
-    tr->set_file_format_version(get_file_format_version());
+    // Need to get a real version_id in case the caller asks for the default/most recent version:
+    bool asked_for_most_recent = version_id == VersionID();
+    grab_read_lock(read_lock, version_id);
+    version_id.version = read_lock.m_version;
+    // reuse existing frozen transaction if one exists:
+    std::weak_ptr<Transaction>& weak_tr = m_frozen_transactions[version_id.version];
+    if (auto res = weak_tr.lock()) {
+        return res; // <-- also releases read_lock - the now shared frozen transaction has its own
+    }
+    // reuse most recent (and add to cache) if asked for it
+    if (asked_for_most_recent && m_recent_frozen_transaction &&
+        m_version_of_recent_frozen_transaction == version_id.version) {
+        weak_tr = m_recent_frozen_transaction;
+        return m_recent_frozen_transaction;
+    }
+    // none cached, create a new.
```

Comment:
`new one.` or `new Transaction.` (or just remove the comment since the next line is clearly creating a `new Transaction` :smile:)